test/zdtm/static/socket_icmp.c

```diff
@@ -0,0 +1,109 @@
+#include "zdtmtst.h"
+
+const char *test_doc = "static test for ICMP socket\n";
+const char *test_author = "समीर सिंह Sameer Singh <lumarzeli30@gmail.com>\n";
+
+/* Description:
+ * Send a ping to localhost using ICMP socket
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <netinet/ip_icmp.h>
+#include <arpa/inet.h>
+#include <sys/time.h>
+#include <netdb.h>
+
+#define PACKET_SIZE 64
+#define RECV_TIMEOUT 1
+
+int ping(void)
+{
+	int ret, sock, seq = 0, recv_len = 0;
+	char packet[PACKET_SIZE], recv_packet[PACKET_SIZE];
+
+	struct timeval tv;
+	struct icmphdr icmp_header, *icmp_reply;
+	struct sockaddr_in addr, recv_addr;
+	socklen_t addr_len;
+
+	sock = socket(PF_INET, SOCK_DGRAM, IPPROTO_ICMP);
+	if (sock < 0) {
+		pr_perror("Can't create socket");
+		return 1;
+	}
+
+
+	tv.tv_sec = RECV_TIMEOUT;
+	tv.tv_usec = 0;
+	if (setsockopt(sock, SOL_SOCKET, SO_RCVTIMEO, &tv, sizeof(tv)) < 0) {
+		pr_perror("Can't set socket option");
+		return 1;
+	}
+
+	memset(&addr, 0, sizeof(addr));
+	addr.sin_family = AF_INET;
+	addr.sin_addr.s_addr = inet_addr("127.0.0.1");
+
+	for (;;) {
+		memset(&icmp_header, 0, sizeof(icmp_header));
+		icmp_header.type = ICMP_ECHO;
+		icmp_header.code = 0;
+		icmp_header.un.echo.id = getpid();
+		icmp_header.un.echo.sequence = seq++;
+
+		memcpy(packet, &icmp_header, sizeof(icmp_header));
+		memset(packet + sizeof(icmp_header), 0xa5,
+			PACKET_SIZE - sizeof(icmp_header));
+
+		ret = sendto(sock, packet, PACKET_SIZE, 0,
+			(struct sockaddr *)&addr, sizeof(addr));
+
+		if (ret < 0) {
+			fail("Can't send");
+			return 1;
+		}
+
+		addr_len = sizeof(recv_addr);
+
+		ret = recvfrom(sock, recv_packet, sizeof(recv_packet), 0,
+			(struct sockaddr*)&recv_addr, &addr_len);
+
+		if (ret < 0) {
+			fail("Can't recv");
+			return 1;
+		}
+
+		icmp_reply = (struct icmphdr *)recv_packet;
+
+		if (icmp_reply->type == ICMP_ECHOREPLY) {
+			printf("%d bytes from %s: icmp_seq=%d\n",
+				recv_len,
+				inet_ntoa(recv_addr.sin_addr),
+				icmp_reply->un.echo.sequence);
+		}
+
+		sleep(1);
+	}
+
+	close(sock);
+}
+
+int main(int argc, char **argv)
+{
+	int ret;
+	test_init(argc, argv);
+
+	test_daemon();
+	test_waitsig();
```

Comment:
@ss141309 `test_daemon()`/`test_waitsig()` should be called when the socket has been created.
The following page provides more information about the ZDTM API: https://criu.org/ZDTM_API

Reply:
Ideally, the test should fail if the patch that adds support for ICMP sockets has not been applied.

Reply:
@rst0git do you have an idea why I am not able to create an ICMP socket in the zdtm test suite?

```c
#include "zdtmtst.h"
#include <unistd.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <netinet/ip_icmp.h>
#include <arpa/inet.h>
#include <sys/time.h>
#include <netdb.h>

int main(int argc, char **argv)
{
  int sock;
  test_init(argc, argv);

  sock = socket(PF_INET, SOCK_DGRAM, IPPROTO_ICMP);
	if (sock < 0) {
		pr_perror("Can't create socket");
		return 1;
	}

	test_daemon();
	test_waitsig();

        close(sock);
	pass();
	return 0;
}
```

I run the test like this and get the output:
```
ss141309@fedora:~/criu/test$ sudo ./zdtm.py run --ignore-taint -a
[sudo] password for ss141309: 
userns is supported
The kernel is tainted: '12289'
Skipping zdtm/lib/groups (manual run only)
Skipping zdtm/static/conntracks (manual run only)
=== Run 3/3 ================ zdtm/static/socket_icmp
====================== Run zdtm/static/socket_icmp in ns =======================
 DEP       socket_icmp.d
 CC        socket_icmp.o
 LINK      socket_icmp
Start test
./socket_icmp --pidfile=socket_icmp.pid --outfile=socket_icmp.out
The test failed (0, 1)
make: *** [Makefile:504: socket_icmp.pid] Error 1
 Test zdtm/static/socket_icmp FAIL at ['make', '--no-print-directory', '-C', 'zdtm/static', 'socket_icmp.pid'] 
Test output: ================================
10:23:46.792:     4: ERR: socket_icmp.c:25: Can't create socket (errno = 13 (Permission denied))
10:23:46.792:     3: ERR: test.c:320: Test exited unexpectedly with code 1

 <<< ================================
##################################### FAIL #####################################
```

Reply:
ICMP sockets probably require CAP_NET_RAW. You need to add the suid flag for this test. Here is an example: https://github.com/checkpoint-restore/criu/blob/criu-dev/test/zdtm/static/socket-raw.desc

Reply:
@avagin using ICMP protocol with SOCK_DGRAM allows it to be run rootless and without requiring CAP_NET_RAW, also even if I add an suid flag in the desc file, the test fails.

Is this a problem with my `net.ipv4.ping_group_range` variable? It is set to '0 2147483647', therefore the program should have no problem running unprivileged ICMP sockets, this is evident when I run the test manually
```
make cleanout
sudo make socket_icmp
make socket_icmp.pid
// Then C-R
make socket_icmp.out
```

The test passes with no problem

Reply:
@avagin I was also able to reproduce the error locally.

`socket(AF_INET, SOCK_DGRAM, IPPROTO_ICMP)` can be used for [non-privileged ICMP](https://manpagez.com/man/4/icmp/). This works when used outside ZDTM. However, the ZDTM environment causes `socket()` to fail with `Permission denied`.

> Is this a problem with my net.ipv4.ping_group_range variable?

The test still fails when `net.ipv4.ping_group_range` is set to `0 2147483647`

Reply:
All zdtm tests are running in separate (new) network namespaces. We have two places where test  namespaces are created:
* test/zdtm_ct.c
* test/zdtm/lib/ns.c

In both cases, we need to set net.ipv4.ping_group_range.  BTW, the net.ipv4.ping_group_range sysctl value should be C/R-ed.